rust/src/lib.rs

```diff
@@ -381,74 +371,72 @@ pub unsafe extern "C" fn zcashlc_derive_extended_spending_keys(
 ///   many bytes, and it must be properly aligned. This means in particular:
 ///   - The entire memory range pointed to by `ptr` must be contained within a single allocated
 ///     object. Slices can never span across multiple allocated objects.
 ///   - `ptr` must be non-null and aligned even for zero-length slices.
 ///   - `ptr` must point to `len` consecutive properly initialized values of type
-///     [`FFIUnifiedViewingKey`].
-/// - The total size `len * mem::size_of::<FFIUnifiedViewingKey>()` of the slice pointed to
+///     [`FFIEncodedKey`].
+/// - The total size `len * mem::size_of::<FFIEncodedKey>()` of the slice pointed to
 ///   by `ptr` must be no larger than isize::MAX. See the safety documentation of pointer::offset.
-/// - See the safety documentation of [`FFIUnifiedViewingKey`]
+/// - See the safety documentation of [`FFIEncodedKey`]
 #[repr(C)]
-pub struct FFIUVKBoxedSlice {
-    ptr: *mut FFIUnifiedViewingKey,
+pub struct FFIEncodedKeys {
+    ptr: *mut FFIEncodedKey,
     len: usize, // number of elems
 }
 
-fn ufvk_to_ffi(
-    account_id: AccountId,
-    ufvk: &UnifiedFullViewingKey,
-    network: Network,
-) -> FFIUnifiedViewingKey {
-    let encoded_ufvk = ufvk.encode(&network);
-
-    FFIUnifiedViewingKey {
-        account_id: account_id.into(),
-        encoding: CString::new(encoded_ufvk).unwrap().into_raw(),
+impl FFIEncodedKeys {
+    pub fn ptr_from_vec(v: Vec<FFIEncodedKey>) -> *mut Self {
+        // Going from Vec<_> to Box<[_]> just drops the (extra) `capacity`
+        let boxed_slice: Box<[FFIEncodedKey]> = v.into_boxed_slice();
+        let len = boxed_slice.len();
+        let fat_ptr: *mut [FFIEncodedKey] = Box::into_raw(boxed_slice);
+        let slim_ptr: *mut FFIEncodedKey = fat_ptr as _;
+        Box::into_raw(Box::new(FFIEncodedKeys { ptr: slim_ptr, len }))
     }
 }
 
-fn ufvk_vec_to_ffi(v: Vec<FFIUnifiedViewingKey>) -> *mut FFIUVKBoxedSlice {
-    // Going from Vec<_> to Box<[_]> just drops the (extra) `capacity`
-    let boxed_slice: Box<[FFIUnifiedViewingKey]> = v.into_boxed_slice();
-    let len = boxed_slice.len();
-    let fat_ptr: *mut [FFIUnifiedViewingKey] = Box::into_raw(boxed_slice);
-    let slim_ptr: *mut FFIUnifiedViewingKey = fat_ptr as _;
-    Box::into_raw(Box::new(FFIUVKBoxedSlice { ptr: slim_ptr, len }))
-}
-
-/// Frees an array of FFIUVKBoxedSlice values as allocated by `zcashlc_derive_unified_viewing_keys_from_seed`
+/// Frees an array of FFIEncodedKeys values as allocated by `zcashlc_derive_unified_viewing_keys_from_seed`
 ///
 /// # Safety
 ///
-/// - `uvks` must be non-null and must point to a struct having the layout of [`FFIUVKBoxedSlice`]
-///   with alignment corresponding to the size of a pointer. See the safety documentation of
-///   [`FFIUVKBoxedSlice`].
+/// - `ptr` must be non-null and must point to a struct having the layout of
+///   [`FFIEncodedKeys`] with alignment corresponding to the size of a pointer. See the
```

Comment:
This should be "`ptr` must be non-null and must point to a correctly-aligned struct having the layout of [`FFIEncodedKeys`]."

The C Standard doesn't place any requirement on the alignment of structs. You can infer that:
* The alignment of a struct must be a multiple of the alignment of each of its non-bitfield members;
* The alignment of any type, including struct types, must divide the alignment provided by general allocation functions such as `malloc`.

This leaves the possibility that struct alignment could be greater than the LCM of the alignments of its non-bitfield members. I don't know of platform ABIs that do that, but it's not excluded in C. So the suggested wording above is correct (and fine in practice because even creating a non-null pointer to a struct that is incorrectly assigned is already undefined behaviour on the C/C++ side).